### PR TITLE
Fix cli output path

### PR DESCRIPTION
**Pull request by leowrites**
## Proposed Changes
Fix the `--output` flag to correctly parse the directory name and the file name.

...

<details>
<summary>Screenshots of your changes (if applicable)</summary>

</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |        |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |          |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |        x  |
| 🎨 _User interface change_ (change to user interface; provide screenshots)              |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

-   [x] I have performed a self-review of my changes.
    -   Check that all changed files included in this pull request are intentional changes.
    -   Check that all changes are relevant to the purpose of this pull request, as described above.
-   [x] I have added tests for my changes, if applicable.
    -   This is **required** for all bug fixes and new features.
-   [x] I have updated the project documentation, if applicable.
    -   This is **required** for new features.
-   [x] If this is my first contribution, I have added myself to the list of contributors.
-   [x] I have updated the project Changelog (this is required for all changes).

After opening your pull request:

-   [x] I have verified that the CI tests have passed.
-   [x] I have reviewed the test coverage changes reported by Coveralls.
-   [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

### CODE REVIEW

File: memory-viz/bin/cli.js [hunk 1]
```diff
@@ -7,26 +7,31 @@ const { program } = require("commander");
...
-    return pathExists(input, `Output path`);
-}
-
-// helper function for parsing paths
-function pathExists(inputPath, errMsg) {
-    const fullPath = path.resolve(process.cwd(), inputPath);
-    if (!fs.existsSync(fullPath)) {
-        console.error(`Error: ${errMsg} ${fullPath} does not exist.`);
+    if (isPathDirectory(input)) {
+        console.error(`Error: Output path must be a file`);
```
Comment (leowrites) [inline memory-viz/bin/cli.js:20]: Missing period

File: memory-viz/bin/cli.js [hunk 2]
```diff
@@ -7,26 +7,31 @@ const { program } = require("commander");
 const { json } = require("node:stream/consumers");
 
 function parseFilePath(input) {
-    if (input) {
-        return pathExists(input, `File`);
-    } else {
-        return undefined;
+    const fullPath = path.resolve(process.cwd(), input);
+    if (fs.existsSync(fullPath, "File")) {
```
Comment (leowrites) [inline memory-viz/bin/cli.js:11]: The input path can really be anything and might not even be syntactically valid. Do we want to do a syntax check before trying to resolve the path?

We can use a regex like this: https://stackoverflow.com/questions/73595422/how-do-i-know-if-a-given-path-is-valid-regardless-of-the-existence-or-absence-o

File: memory-viz/bin/cli.js [hunk 3]
```diff
@@ -7,26 +7,31 @@ const { program } = require("commander");
...
         process.exit(1);
-    } else {
-        return fullPath;
     }
+    const parsedPath = path.parse(input);
+    const outputDir = parsedPath.dir || ".";
+    const filename = `${parsedPath.name}.svg`;
+    if (!fs.existsSync(outputDir)) {
+        console.error(`Error: Output directory ${outputDir} does not exist`);
+        process.exit(1);
```
Comment (leowrites) [inline memory-viz/bin/cli.js:28]: This code is reused in so many places that it should have its own function

File: memory-viz/bin/cli.js [hunk 4]
```diff
@@ -7,26 +7,31 @@ const { program } = require("commander");
...
+    const filename = `${parsedPath.name}.svg`;
+    if (!fs.existsSync(outputDir)) {
+        console.error(`Error: Output directory ${outputDir} does not exist.`);
+        process.exit(1);
+    }
+    return path.join(outputDir, filename);
+}
+
+function isPathDirectory(rawOutputPath) {
+    return rawOutputPath.slice(-1) == "/";
```
Comment (leowrites) [inline memory-viz/bin/cli.js:34]: Use "`===`

File: memory-viz/bin/cli.js [hunk 5]
```diff
@@ -7,26 +7,31 @@ const { program } = require("commander");
...
-        console.error(`Error: ${errMsg} ${fullPath} does not exist.`);
+    if (isPathDirectory(input)) {
+        console.error(`Error: Output path ${input} must be a file.`);
         process.exit(1);
-    } else {
-        return fullPath;
     }
+    const parsedPath = path.parse(input);
+    const outputDir = parsedPath.dir || ".";
+    const filename = `${parsedPath.name}.svg`;
```
Comment (david-yz-liu) [inline memory-viz/bin/cli.js:25]: Use `parsedPath.base` to preserve the full filename the user inputs; we don't want to force a specific extension that overwrites whatever the user put in.

File: memory-viz/bin/cli.js [hunk 6]
```diff
@@ -7,26 +7,31 @@ const { program } = require("commander");
...
+    if (!fs.existsSync(outputDir)) {
+        console.error(`Error: Output directory ${outputDir} does not exist.`);
+        process.exit(1);
+    }
+    return path.join(outputDir, filename);
+}
+
+function isPathDirectory(rawOutputPath) {
```
Comment (david-yz-liu) [inline memory-viz/bin/cli.js:33]: This function isn't quite robust enough; consider, for example, someone who puts in a path `./hello` where `hello` is an existing directory.

You should figure out how to use Node's built-in functions to to tell whether an existing path refers to a directory or not.

File: memory-viz/src/tests/cli.spec.tsx [hunk 7]
```diff
@@ -64,11 +49,9 @@ describe.each([
         exec(`memory-viz ${command}`, (err) => {
             if (err) throw err;
 
-            const svgFilePath = getSVGPath(command.includes("--output"));
-            const fileContent = fs.readFileSync(svgFilePath, "utf8");
-
+            const fileContent = fs.readFileSync(outputPath, "utf8");
```
Comment (david-yz-liu) [inline memory-viz/src/tests/cli.spec.tsx:52]: It's more robust to keep the `.includes("--output")`: this does not assume that `--output` is always set to `outputPath`.

File: memory-viz/bin/cli.js [hunk 8]
```diff
@@ -7,26 +7,32 @@ const { program } = require("commander");
 const { json } = require("node:stream/consumers");
 
 function parseFilePath(input) {
-    if (input) {
-        return pathExists(input, `File`);
-    } else {
-        return undefined;
+    const fullPath = path.resolve(process.cwd(), input);
+    if (fs.existsSync(fullPath, "File")) {
```
Comment (leowrites) [inline memory-viz/bin/cli.js:11]: Remove "File"

File: memory-viz/bin/cli.js [hunk 9]
```diff
@@ -7,26 +7,32 @@ const { program } = require("commander");
...
 function parseOutputPath(input) {
-    return pathExists(input, `Output path`);
-}
-
-// helper function for parsing paths
-function pathExists(inputPath, errMsg) {
-    const fullPath = path.resolve(process.cwd(), inputPath);
-    if (!fs.existsSync(fullPath)) {
-        console.error(`Error: ${errMsg} ${fullPath} does not exist.`);
+    if (fs.existsSync(input) && fs.lstatSync(input).isDirectory()) {
```
Comment (leowrites) [inline memory-viz/bin/cli.js:19]: Determine when to use `lstatSync` and `statSync`

**Review by david-yz-liu — APPROVED** [review]
@leowrites great work!

**Pull request merged**